src/bson_mapper/bson_archiver.hpp

```diff
@@ -0,0 +1,1078 @@
+     */
+    void readNextDoc() {
+        // Determine the size of the BSON document in bytes.
+        // TODO: Only works on little endian.
+        int32_t docsize;
+        char docsize_buf[sizeof(docsize)];
+        _readStream.read(docsize_buf, sizeof(docsize));
+        std::memcpy(&docsize, docsize_buf, sizeof(docsize));
+
+        // Throw an exception if the end of the stream is prematurely reached.
+        if (_readStream.eof() || !_readStream) {
+            throw cereal::Exception("No more data in BSONInputArchive stream.\n");
+        }
+
+        // Create a buffer to store the BSON document.
+        char* bsonData = new char[docsize];
+
+        // Read the BSON data from the stream into the buffer.
+        _readStream.seekg(-sizeof(docsize), std::ios_base::cur);
+        _readStream.read(bsonData, docsize);
+
+        // Make sure there were no errors reading the BSON data.
+        if (_readStream.eof() || !_readStream) {
+            throw cereal::Exception("No more data in BSONInputArchive stream.\n");
+        }
+
+        // Store the BSON data of the document in our container of documents.
+        _curBsonDoc =
+            bsoncxx::document::value{reinterpret_cast<uint8_t*>(bsonData),
+                                     static_cast<size_t>(docsize), [](uint8_t* p) { delete[] p; }};
+
+        // Specify that we've read a document.
+        _readFirstDoc = true;
+    }
+
+    /**
+     * Searches for the next BSON element to be retrieved and loaded.
+     *
+     * @return the element with _nextName as a key, or if the current node is an
+     *         array, the next element in that array.
+     */
+    inline bsoncxx::types::value search() {
+        // Make sure the node stack is not empty.
+        if (_nodeTypeStack.empty()) {
+            throw cereal::Exception("Cannot search for element if not in node.\n");
+        }
+
+        // Check to make sure a document has been read.
+        if (!_readFirstDoc) {
+            throw cereal::Exception("Error in element search, never loaded BSON data from file.\n");
+        }
+
+        if (_nextName) {
+            // Reset _nextName
+            std::string nextName = _nextName;
+            _nextName = nullptr;
+
+            // If we're in an object in the Root (InObject),
+            // look for the key in the current BSON view.
+            if (_nodeTypeStack.top() == InputNodeType::InObject ||
+                _nodeTypeStack.top() == InputNodeType::InRootElement) {
+                auto elemFromDoc = _curBsonDoc.view()[nextName];
+
+                if (elemFromDoc) {
+                    return elemFromDoc.get_value();
+                }
+            }
+
+            // If we're in an embedded object, look for the key in the object
+            // at the top of the embedded object stack.
+            if (_nodeTypeStack.top() == InputNodeType::InEmbeddedObject) {
+                auto elemFromDoc = _embeddedBsonDocStack.top()[nextName];
```

Comment:
Done here, and for the other elemFromDoc.

For the elemFromArr I only did "const auto" because if I did const auto&, the reference would point to a different element after incrementing the array iterator.